### ISSUE TITLE
Min and Max with complex inputs exhibit behavior incompatible with NumPy

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#36374]
As the title says, PyTorch's min and max are incompatible with NumPy's. For example:

```
a = np.array((0 + 4j, 4 + 0j, -2 - 2j, 1 + 1j, 2 + 2j, 3 + 3j))
t = torch.from_numpy(a)

np_result = np.min(a)
torch_result = torch.min(t)
```

np_result is (-2 - 2j) while torch_result is (1. + 1.j). The latter seems hard to justify under any notion of min.

Max is no better:

```
np_result = np.max(a)
torch_result = torch.max(t)
```

np_result is (4 + 0j) while torch_result is (-inf + 0.j). Again, the torch_result seems hard to justify under any notion of max.

When a dim is supplied to max the PyTorch behavior changes, but is still incompatible with NumPy's:

```
torch_result = torch.max(t, dim=0)
```

Here torch_result is (3. + 3.j), which is at least a reasonable max value.

cc @ezyang @anjali411 @dylanbespalko

**Comment by n-west** [comment]
Hi, this is because the minimum and maximum definition in vec256_complex_{double,float}.h are using the abs (really the square abs) to find the minimum and maximum. Numpy uses "The sort order for complex numbers is lexicographic. If both the real and imaginary parts are non-nan then the order is determined by the real parts except when they are equal, in which case the order is determined by the imaginary parts." (https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.sort.html)

I'm a longish-time user of torch, interested in complex support, have never had to dig these deep yet but was perusing issues to see how I might be able to help out. I'm happy to fix this

1) would you like the same sort order as numpy or would you prefer another definition of minimum/maximum sorting?
2) For the minimum and maximum definitions in vec256 files, is there a possible side effect outside of these pytorch ops? Like is there a potential caller of this that is expecting the square-abs such that a new function should be defined and used by the min/max ops?

**Comment by dylanbespalko** [comment]
@n-west 

Thanks for helping. I wrote the initial code that you are referring to in the vec256 files. At the time, the maximum and minimum functions were not callable by the old TH library. Therefore, I don't think there is any downstream code yet, atleast not for complex numbers.

#33568 addresses a related issue with `torch.clamp` vs. `numpy.clip`. Using `abs()` to select the highest value is the logical solution, **however the easily programmable and numpy compatible solution might initially be preferred.**

To delve deeper, the C++ standard library doesn't define less-than or greater-than operators for complex numbers. If it is not in the spec, then it won't be accelerated on the GPU, ... #35524 is redefining the `std::complex<T>` within PyTorch to simplify CPU/GPU interoperability. This would also allow you to define a less-than and greater-than operator directly in `std::complex<T>` and make it easier to define clamp and min/max operations using the `abs` method.

**Comment by mruberry** [comment]
Hi @n-west, thanks for your interest! And thanks for your comment, too, @dylanbespalko.

I'd like to think about this some more, but I actually wonder if PyTorch should allow comparisons between complex numbers at all? 

On one hand, the complex numbers aren't part of any totally ordered field, and neither Python nor C++ support direct comparisons between complex numbers. This approach also gives us time to identify users who want a particular kind of comparison so we can review how best to meet their needs.

On the other hand, a lexicographic ordering of complex values, like NumPy uses, is intuitive.

Maybe that should be the discussion we're having (in a separate issue)? If we don't support a "default" comparator between complex values then, logically, we shouldn't support min, max, clamp, or sort for complex tensors, either, since they're comparison-based. Although in the future we could allow users to specify their own comparator, like you can with std::sort.

**Comment by n-west** [comment]
I agree with using abs() or equivalently L2-norm is the logical thing (at least in the applications I work with... I'm not sure when I'd ever want to sort the numpy way). As you both mentioned C++ and most other languages don't define these comparisons at all because there's not necessarily a mathematical definition I know of for the comparison.

For what it's worth I'm coming from complex baseband radio processing. I've had users complain about similar things for code I have maintained (https://github.com/gnuradio/volk/issues/139), but I think it's mostly because it's a subtly confusing function for complex numbers and the lexigraphic approach isn't necessarily intuitive either. I'll hold off for now and maybe look for some other complex support I can toss in. I'm pretty excited to see it coming!

**Comment by mruberry** [comment]
Thanks for that context, @n-west! And glad to hear you're excited. We are, too!

Did you know I recently updated torch.abs() to return a float tensor when given a complex input? Would a combination of torch.abs(), argsort, and index_select work to implement sorting by absolute value? We could also think of implementing a function like "sort_absolute" that sorted any array using absolute value to compare numbers.

**Comment by n-west** [comment]
Yes, those would be useful. I don't necessarily care about sorting most of the time, but I do care about min and max. Currently I do the following:
```
        auto samples_re = samples.slice(-1, 0, 1);
        auto samples_im = samples.slice(-1, 1, 2);
        auto samples_maxabs = std::get<0>(torch::max(samples_re * samples_re + samples_im * samples_im, {1,}, true));
        samples = samples / (_eps + samples_maxabs);
```

I believe the existing `torch::norm` can replace the slice and multiplies, probably more efficiently. Having a proper complex number is mostly icing on the cake.

One thing that may be useful context is there's a growing world of people running neural networks on complex radio samples. It's not the most rigorous thing in the world, but it's extremely popular to feed IQ as the initial feature dimension to neural networks (see https://arxiv.org/abs/1602.04105 & https://arxiv.org/abs/1703.09197 & https://arxiv.org/pdf/1702.00832.pdf). There's been some work on proper complex convolutions to preserve phase. The community is still split on whether that helps anything or not (see III.C.1 of https://arxiv.org/pdf/1605.00716.pdf)... That being said, I haven't looked at what conversion routines between the two forms might be. That community would benefit from something like the numpy approach cat(array.real, array.imag), but I'm getting a bit off topic here :-)

**Comment by mruberry** [comment]
Let's move this discussion to the more general discussion in https://github.com/pytorch/pytorch/issues/36444.

**Comment by ezyang** [comment]
also cc @rgommers

**Comment by mruberry** [comment]
Following discussion in #36444, it seems we have consensus on disabling min and max for now.
